File: Depth-First Search/Backtracking/Medium/LET_no1286_Iterator for Combination.cpp
class CombinationIterator {
private:
    vector<int> idx;
    string s;
public:
    CombinationIterator(string characters, int combinationLength) {
        s = characters;
        for (int i = 0; i < combinationLength; i++)
            idx.push_back(i);
    }
    
    string next() {
        string res;
        for (int i = 0; i < idx.size(); i++)
            res += s[idx[i]];
        
        idx.back()++;
        if (idx.back() < s.size())
            return res;
        
        for (int i = idx.size() - 2, remain = 2; i >= 0; i--, remain++) {
            if (idx[i] + remain < s.size()) {
                idx[i]++;
                for (int j = i + 1; j < idx.size(); j++)
                    idx[j] = idx[j - 1] + 1;
                return res;
            }
        }
        
        idx[0] = -1;
        return res;
    }
    
    bool hasNext() {
        return idx[0] != -1;
    }
};

/**
 * Your CombinationIterator object will be instantiated and called as such:
 * CombinationIterator* obj = new CombinationIterator(characters, combinationLength);
 * string param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

File: Depth-First Search/Backtracking/Medium/LET_no1286_Iterator for Combination.py
class CombinationIterator:

    def __init__(self, characters: str, combinationLength: int):
        self.s = characters
        self.idx = [i for i in range(combinationLength)]

    def next(self) -> str:
        res = ""
        for i in self.idx:
            res += self.s[i]
        
        self.idx[-1] += 1
        if self.idx[-1] < len(self.s):
            return res
        
        i, remain = len(self.idx) - 2, 2
        while i >= 0:
            if self.idx[i] + remain < len(self.s):
                self.idx[i] += 1
                for j in range(i + 1, len(self.idx)):
                    self.idx[j] = self.idx[j - 1] + 1
                return res
            i -= 1
            remain += 1
        
        self.idx[0] = -1
        return res

    def hasNext(self) -> bool:
        return self.idx[0] != -1


# Your CombinationIterator object will be instantiated and called as such:
# obj = CombinationIterator(characters, combinationLength)
# param_1 = obj.next()
# param_2 = obj.hasNext()